### PR TITLE
Refactoring validation middleware

### PR DESCRIPTION
**Pull request by roschaefer**
## Pullrequest
Two things:

* Fix `UpdateUser` test block (was not testing anything).
* Put all validations in one place

@mattwr18 @ulfgebhardt please read my commit messages in this PR.
### Issues
This is a follow-up on #478 

### Checklist

- [X] None

### How2Test

- [X] None

### Todo

- [X] None

### CODE REVIEW

File: backend/src/resolvers/users.spec.js [hunk 1]
```diff
@@ -5,15 +5,15 @@ import Factory from '../seed/factories'
 const factory = Factory()
 let client
 
-afterAll(async () => {
+afterEach(async () => {
```
Comment (roschaefer) [inline backend/src/resolvers/users.spec.js]: @ulfgebhardt `afterEach` means: Run this after each test case. `afterAll` means: Run this after the entire test **suite**. `afterAll` is great for performance tweaks, but you have to tear down the created data yourself. Read: https://jestjs.io/docs/en/setup-teardown

File: backend/src/middleware/validation/index.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
+  } else {
+    throw new UserInputError('Input is not a URL')
+  }
+}
+
+export default {
+  Mutation: {
+    CreateUser: validateUsername,
```
Comment (roschaefer) [inline backend/src/middleware/validation/index.js]: @ulfgebhardt we should validate `username` also on creation, no? Without it, the validation would be kind of pointless.

... Except, the `CreateUser` will be called on registration and we supply no username at all. But we'll see. For now, I think it's safer to have it checked on Create+Update.

File: backend/src/resolvers/users.spec.js [hunk 3]
```diff
@@ -35,34 +36,39 @@ describe('userMiddleware', () => {
     })
   })
 
-  describe('update User', () => {
+  describe('UpdateUser', () => {
+    beforeEach(async () => {
+      await factory.create('User', { id: 'u47', name: 'John Doe' })
```
Comment (roschaefer) [inline backend/src/resolvers/users.spec.js]: @ulfgebhardt you tried to update a user that did not exist.

**Comment by roschaefer** [comment]
Me: Adding/fixing tests
Codecov: Test coverage dropped by 0.01%.

:laughing:

File: backend/src/middleware/validation/index.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { UserInputError } from 'apollo-server'
+
+const USERNAME_MIN_LENGTH = 3
+
+const validateUsername = async (resolve, root, args, context, info) => {
+  if (args.name && args.name.length >= USERNAME_MIN_LENGTH) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
```
Comment (mattwr18) [inline backend/src/middleware/validation/index.js]: are we updating the user twice here? once in the validation middleware and once on line 15 of the `userMiddleware`?

File: backend/src/middleware/validation/index.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    throw new UserInputError(`Username must be at least ${USERNAME_MIN_LENGTH} characters long!`)
+  }
+}
+
+const validateUrl = async (resolve, root, args, context, info) => {
+  const { url } = args
+  const isValid = url.match(/^(?:https?:\/\/)(?:[^@\n])?(?:www\.)?([^:/\n?]+)/g)
+  if (isValid) {
+    /* eslint-disable-next-line no-return-await */
+    return await resolve(root, args, context, info)
```
Comment (mattwr18) [inline backend/src/middleware/validation/index.js]: same as :point_up:

File: backend/src/middleware/validation/index.js (hunk 4 above)
Comment (roschaefer) [reply backend/src/middleware/validation/index.js]: No, calling `resolve` will continue the chain of middlewares until you hit the final resolver.

File: backend/src/resolvers/users.spec.js [hunk 6]
```diff
@@ -71,7 +77,8 @@ describe('userMiddleware', () => {
 
     it('with too short name', async () => {
       const variables = {
-        id: 'u1'
+        id: 'u47',
```
Comment (roschaefer) [inline backend/src/resolvers/users.spec.js]: I think I have to give myself a "Request Changes" review here. I told @ulfgebhardt not to create test cases (== specification) for behaviour that we actually don't want to support, e.g. passing `id` from the client. I'll update that as soon as possible.

**Comment by ulfgebhardt** [comment]
> Me: Adding/fixing tests
> Codecov: Test coverage dropped by 0.01%.
> 
> 😆

The Backend Test Setup needs to be fixed to comply with standard test runs of jest in oder to collect coverage properly.

**Comment by roschaefer** [comment]
@ulfgebhardt did you see this? Check commit messages and "Files Changed" tab. If yes, you can merge this PR.

**Pull request merged**